apis/monitoring/v1beta1/alertpolicy_reference.go

```diff
@@ -95,3 +95,43 @@ func (r *MonitoringAlertPolicyRef) NormalizedExternal(ctx context.Context, reade
 	r.External = name
 	return r.External, nil
 }
+
+var _ refsv1beta1.RefinerWithProjectID = &MonitoringAlertPolicyRef{}
+
+func (r *MonitoringAlertPolicyRef) RefineWithProjectID(projectID, path string) error {
+	tokens := strings.Split(r.External, "/")
+
+	// AlertChart has a specific format requirement:
+	// "The format is: projects/[PROJECT_ID_OR_NUMBER]/alertPolicies/[ALERT_POLICY_ID]"
+	if strings.HasSuffix(path, ".AlertChart.AlertPolicyRef") {
+		// alertPolicies/{{alertPolicyId}}
+		if len(tokens) == 2 && tokens[0] == "alertPolicies" {
+			r.External = fmt.Sprintf("projects/%s/%s", projectID, r.External)
+			return nil
+		}
+		// projects/{{projectId}}/alertPolicies/{{alertPolicyId}}
+		if tokens[1] != projectID {
+			return fmt.Errorf("resolve alertPolicy (%q) in AlertChat was not in same project", r.External)
+		}
+		return nil
+	}
+
+	// IncidentList has a specific format requirement:
+	// "Don't include the project ID prefix in the policy name. For example, use `alertPolicies/utilization`."
+	if strings.HasSuffix(path, ".IncidentList.AlertPolicyRef") {
```

Comment:
Can we fix this in the mapper instead?

Reply:
I think normalization should take it to normalized form (only external set, in a format we define, including project id vs number).  Then if that isn't the format the API expects, we transform that in the controller or the mapper.

I think either we visit the fields in the controller before calling the mapper, or we embed it into the mapper.  You're right, the mapper is probably not a good place because the mapper is generated.